File: Data Structures & Algorithms/count-number-of-islands/submission-2.java
class Solution {
    public char LAND = '1';
    public char WATER = '0';
    public int m;
    public int n;

    public int numIslands(char[][] grid) {
        n = grid.length;
        m = grid[0].length;
        int count = 0;

        for (int r = 0; r < n; r++) {
            for (int c = 0; c < m; c++) {
                if(grid[r][c] == LAND){
                    count++;
                    dfs(grid, r, c);
                }
            }
        }

        return count;
    }

    public void dfs(char[][] grid, int r, int c) {
        if (isValid(grid, r, c)) {
            grid[r][c] = WATER;
            dfs(grid, r + 1, c);
            dfs(grid, r - 1, c);
            dfs(grid, r, c + 1);
            dfs(grid, r, c - 1);
        }
    }

    public boolean isValid(char[][] grid, int r, int c) {
        if (Math.min(r, c) < 0 || r >= n || c >= m || grid[r][c] == WATER) {
            return false;
        }
        return true;
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-2.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        directions = [(0,1),(1,0),(-1,0),(0,-1)]
        m = len(grid)
        n = len(grid[0])
        max_area = 0
        LAND = 1
        WATER = 0

        def isValid(r: int, c: int):
            if min(r, c) < 0 or r >= m or c >= n or grid[r][c] == WATER:
                return False
            return True

        def dfs(r: int, c: int):
            area = 1
            grid[r][c] = WATER

            for dx, dy in directions:
                x = r + dx
                y = c + dy

                if isValid(x, y):
                    area += dfs(x, y)

            return area

        for r in range(m):
            for c in range(n):
                if grid[r][c] == LAND:
                    max_area = max(max_area, dfs(r, c))

        return max_area